serverless/internal/storage/webstorage/fs.go

```diff
@@ -0,0 +1,290 @@
+	"github.com/google/trillian-examples/serverless/api/layout"
+	"github.com/google/trillian-examples/serverless/internal/storage"
+)
+
+// Storage is a serverless storage implementation which uses webstorage entries to store tree state.
+// The storage key format is:
+//  <root>/leaves/aa/bb/cc/ddeeff...
+//  <root>/leaves/pending/aabbccddeeff...
+//  <root>/seq/aa/bb/cc/ddeeff...
+//  <root>/tile/<level>/aa/bb/ccddee...
+//  <root>/checkpoint
+//
+// The functions on this struct are not thread-safe.
+type Storage struct {
+	// root is the prefix of all webstorage keys which store tree data.
+	root string
+	// nextSeq is a hint to the Sequence func as to what the next available
+	// sequence number is to help performance.
+	// Note that nextSeq may be <= than the actual next available number, but
+	// never greater.
+	nextSeq uint64
+	// checkpoint is the latest known checkpoint of the log.
+	checkpoint log.Checkpoint
+}
+
+const leavesPendingPathFmt = "leaves/pending/%0x"
+
+func getStorage() js.Value {
+	return js.Global().Get("sessionStorage")
+}
+
+func exists(k string) bool {
+	_, err := get(k)
+	return err == nil
+}
+
+func get(k string) ([]byte, error) {
+	v := getStorage().Call("getItem", js.ValueOf(k))
+	if js.Null().Equal(v) {
+		return nil, os.ErrNotExist
+	}
+	return []byte(v.String()), nil
+}
+
+func set(k string, v []byte) error {
+	getStorage().Call("setItem", js.ValueOf(k), js.ValueOf(string(v)))
+	// TODO(al) read back and check
+	return nil
+}
+
+func createExclusive(k string, v []byte) error {
+	if exists(k) {
+		return os.ErrExist
+	}
+	getStorage().Call("setItem", js.ValueOf(k), js.ValueOf(string(v)))
+	// TODO(al) read back and check
+	return nil
+}
+
+// Load returns a Storage instance initialised from webstorage prefixed at root.
+func Load(root string, checkpoint *log.Checkpoint) (*Storage, error) {
+	return &Storage{
+		root:       root,
+		checkpoint: *checkpoint,
+		nextSeq:    checkpoint.Size,
+	}, nil
+}
+
+// Create creates a new filesystem hierarchy and returns a Storage representation for it.
+func Create(root string, emptyHash []byte) (*Storage, error) {
+	fs := &Storage{
+		root:    root,
+		nextSeq: 0,
+		checkpoint: log.Checkpoint{
+			Size: 0,
+			Hash: emptyHash,
+		},
+	}
+
+	return fs, nil
+}
+
+// Queue adds a leaf to the pending queue for integration.
+func (fs *Storage) Queue(leaf []byte) error {
+	h := sha256.Sum256(leaf)
+	return createExclusive(filepath.Join(fs.root, fmt.Sprintf(leavesPendingPathFmt, h)), leaf)
+}
+
+// PendingKeys returns the storage keys associated with pending leaves.
+func (fs *Storage) PendingKeys() ([]string, error) {
+	pk := []string{}
+	s := getStorage()
+	prefix := filepath.Join(fs.root, "leaves", "pending")
+	for i := 0; i < s.Get("length").Int(); i++ {
+		k := s.Call("key", js.ValueOf(i))
+		if js.Null().Equal(k) {
+			return nil, fmt.Errorf("key(%d) failed", i)
+		}
+		ks := k.String()
+		if strings.HasPrefix(ks, prefix) {
```

Comment:
It was only at this point that I realized there isn't even a pretense in the `s // getStorage` variable of a filesystem. It seems to be a collection of key / values, where the keys are structured such that by checking the prefix you can emulate a filesystem. I think that would be worth outlining at a high level in some documentation up there ^.

Reply:
Yup, that's right - it is a key value store.
I don't really want to duplicate the documentation of the WebStorage API, but I've added a link to it in the comment about browsers at the top.